src/app/test/role_test.py

```diff
@@ -0,0 +1,114 @@
+        {
+            "name": "testrole1",
+            "description": "a test role",
+            "permissions": ["garden:read", "garden:create", "garden:update"],
+        },
+        {
+            "name": "testrole2",
+            "permissions": ["garden:read", "garden:delete"],
+        },
+    ]
+
+    yield role_data
+
+
+@pytest.fixture
+def role_sync_data_missing_fields():
+    role_data = [
+        {
+            "name": "badrole",
+            "description": "hey",
+        },
+    ]
+
+    yield role_data
+
+
+@pytest.fixture
+def user_with_role_assignments():
+    role = Role(name="assignedrole1", permissions=["garden:read"]).save()
+    role_assignment = RoleAssignment(domain="garden1", role=role)
+    user = User(username="testuser", role_assignments=[role_assignment]).save()
+
+    yield user
+
+
+class TestRole:
+    @classmethod
+    def setup_class(cls):
+        connect("beer_garden", host="mongomock://localhost")
+        config._CONFIG = {"garden": {"name": "localgarden"}}
+
+    def test_sync_roles_creates_new_roles(self, role_sync_data):
+        """sync_roles should create new roles when none exists"""
+        role_names = [role["name"] for role in role_sync_data]
+
+        sync_roles(role_sync_data)
+
+        assert len(Role.objects.filter(name__in=role_names)) == len(role_sync_data)
+
+    def test_sync_roles_deletes_roles_not_present(self, role_sync_data):
+        """sync_roles should delete roles in the database that are are not present in
+        the role_sync_data
+        """
+        Role(name="notinsyncdata", permissions=["garden:read"]).save()
+
+        sync_roles(role_sync_data)
+
+        with pytest.raises(DoesNotExist):
+            Role.objects.get(name="notinsyncdata")
+
+    def test_sync_roles_updates_existing_roles(self, role_sync_data):
+        """sync_roles should update any existing Roles based on what is in the provided
+        role_sync_data
+        """
+        Role(
+            name="testrole1", description="somethingwitty", permissions=["system:read"]
+        ).save()
+
+        sync_roles(role_sync_data)
+
+        role = Role.objects.get(name="testrole1")
+
+        assert role.description == role_sync_data[0]["description"]
+        assert role.permissions == role_sync_data[0]["permissions"]
+
+    def test_sync_roles_raises_validation_error_for_missing_fields(
+        self, role_sync_data_missing_fields
+    ):
+        """sync_roles should raise a ValidationError if any required fields are missing
+        in the supply input sync data
+        """
+        with pytest.raises(ValidationError):
+            sync_roles(role_sync_data_missing_fields)
+
+    def test_sync_roles_removes_user_role_assignments(
+        self, role_sync_data, user_with_role_assignments
+    ):
+        """sync_roles should remove role assignments references Roles that are deleted
+        as part of the sync process
+        """
+        assert len(user_with_role_assignments.role_assignments) > 0
```

Comment:
I'm curious on opinions here.  It occurs to me that a lot of tests make assumptions about the state of the data before the test runs.  So in this case, my assertion at the end is:

```python
assert len(user_with_role_assignments.role_assignments) == 0
```

With **just** that assertion, someone could make a change to the test data such that `len(user_with_role_assignments.role_assignments)` starts at 0 and this test would continue to pass, even though it was no longer testing anything meaningful.  So I opted to put this upfront assertion to effectively validate that the data is in a state that makes my test meaningful, i.e. this test would now fail if there were no role_assignments to begin with, which would let us know that we've invalidated this test.

Reply:
It sounds like a better idea to be explicit with exactly what behavior you are testing. Tests that have different assumptions about the data being tested should probably have different fixtures, or should modify the fixture in their own namespace. This assertion doesn't seem out of the ordinary given its function.

Reply:
I think this is a fine solution. `pytest` also supports xunit style, method-level setup and teardown, but that would be overkill in this situation IMO.